Modules/_pickle.c

```diff
@@ -6229,6 +6277,39 @@ load_build(UnpicklerObject *self)
         Py_INCREF(slotstate);
         Py_DECREF(tmp);
     }
+    /* proto 5 addition: state can embed a callable state setter */
+    else if (PyTuple_Check(state) && PyTuple_GET_SIZE(state) == 3) {
+        /* borrowed reference*/
+        PyObject *tmp = state;
+
+        setstate = PyTuple_GET_ITEM(tmp, 0);
+        state = PyTuple_GET_ITEM(tmp, 1);
+        slotstate = PyTuple_GET_ITEM(tmp, 2);
+
+        Py_INCREF(setstate);
+        Py_INCREF(state);
+        Py_INCREF(slotstate);
+        Py_DECREF(tmp);
+        /* call the setstate function */
+        if (slotstate != Py_None) {
+            statearg = Py_BuildValue("(OO)", state, slotstate);
+        }
+        else {
+            statearg = state;
+            Py_INCREF(state);
+        }
+
+        if (PyObject_CallFunctionObjArgs(setstate, inst, statearg,
+                                         NULL) == NULL){
+            return -1;
```

Comment:
You probably need to:

```
        Py_DECREF(setstate);
        Py_DECREF(state);
        Py_DECREF(statearg);
        Py_DECREF(slotstate);
```

here no?

Maybe `Py_DECREF(state)` and `Py_DECREF(slotstate)` should be moved up.